src/utils/useTodos.ts

```diff
@@ -0,0 +1,169 @@
+import { useEffect, useState } from 'react';
+import { getTodos } from '../api/todos';
+import { Todo } from '../types/Todo';
+import { User } from '../types/User';
+import { FilterStatus } from '../types/ToDoFilterProps';
+
+export const useTodos = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [filteredTodos, setFilteredTodos] = useState<Todo[]>([]);
+  const [isLoading, setIsLoading] = useState<boolean>(true);
+  const [selectedTodo, setSelectedTodo] = useState<Todo | null>(null);
+  const [selectedUser, setSelectedUser] = useState<User | null>(null);
+  const [isLoadingUser, setIsLoadingUser] = useState<boolean>(false);
+  const [filterStatus, setFilterStatus] = useState<FilterStatus>('all');
+  const [query, setQuery] = useState<string>('');
+  const [selectedTodoId, setSelectedTodoId] = useState<number | null>(null);
+  const [isHovered, setIsHovered] = useState(false);
+  const [errorMessage, setErrorMessage] = useState<string | null>(null);
+
+  // функція для встановлення ерорів
+  const showError = (type: 'load' | 'add' | 'delete' | 'update' | 'empty') => {
+    const messages = {
+      load: 'Unable to load todos',
+      add: 'Unable to add a todo',
+      delete: 'Unable to delete a todo',
+      update: 'Unable to update a todo',
+      empty: 'Title should not be empty',
+    };
+
+    setErrorMessage(messages[type]);
+
+    setTimeout(() => {
+      setErrorMessage(null);
+    }, 3000);
+  };
+
+  // завантаження тудушок з апі
+  useEffect(() => {
+    setIsLoading(true);
+    getTodos()
+      .then((loadedTodos: Todo[]) => {
+        setTodos(loadedTodos);
+        setFilteredTodos(loadedTodos);
+        setErrorMessage(null);
+      })
+      .catch(() => {
+        showError('load');
+      })
+      .finally(() => setIsLoading(false));
+  }, []);
+
+  // фільтрація тудушок при зміні тудушок, значення, відфільтрованого стану
+  useEffect(() => {
+    let result = [...todos];
+
+    if (query) {
+      result = result.filter(todo =>
+        todo.title.toLowerCase().includes(query.toLowerCase()),
+      );
+    }
+
+    if (filterStatus === 'active') {
+      result = result.filter(todo => !todo.completed);
+    } else if (filterStatus === 'completed') {
+      result = result.filter(todo => todo.completed);
+    }
+
+    // console.log('Filter status:', filterStatus);
+    // console.log('Query:', query);
+    // console.log('Todos count:', todos.length);
+    // console.log('Filtered todos count:', result.length);
+
+    setFilteredTodos(result);
```

Comment:
- You don't need additional state `filteredTodos` just store it in a variable. Don't filter in `useEffect`
- Create a helper function to filter and move to utile/helper file/folder
```const filteredTodos = filterTodos(todos, { status });```